src/connector/src/source/cdc/external/sql_server.rs

```diff
@@ -0,0 +1,480 @@
+
+impl SqlServerExternalTable {
+    pub async fn connect(config: ExternalTableConfig) -> ConnectorResult<Self> {
+        tracing::debug!("connect to sql server");
+
+        let mut client_config = Config::new();
+
+        client_config.host(&config.host);
+        client_config.database(&config.database);
+        client_config.port(config.port.parse::<u16>().unwrap());
+        client_config.authentication(tiberius::AuthMethod::sql_server(
+            &config.username,
+            &config.password,
+        ));
+        client_config.trust_cert();
+
+        let tcp = TcpStream::connect(client_config.get_addr()).await?;
+        tcp.set_nodelay(true)?;
+
+        let mut client: tiberius::Client<Compat<TcpStream>> =
+            match tiberius::Client::connect(client_config, tcp.compat_write()).await {
+                // Connection successful.
+                Ok(client) => Ok(client),
+                // The server wants us to redirect to a different address
+                Err(Error::Routing { host, port }) => {
+                    let mut config = Config::new();
+
+                    config.host(&host);
+                    config.port(port);
+                    config
+                        .authentication(tiberius::AuthMethod::sql_server("sa", "YourPassword123"));
+
+                    let tcp = TcpStream::connect(config.get_addr()).await?;
+                    tcp.set_nodelay(true)?;
+
+                    // we should not have more than one redirect, so we'll short-circuit here.
+                    tiberius::Client::connect(config, tcp.compat_write()).await
+                }
+                Err(e) => Err(e),
+            }?;
+
+        let mut column_descs = vec![];
+        let mut pk_names = vec![];
+        {
+            let sql = Query::new(format!(
+                "SELECT * FROM {} WHERE 1 = 0",
+                SqlServerExternalTableReader::get_normalized_table_name(&SchemaTableName {
+                    schema_name: config.schema.clone(),
+                    table_name: config.table.clone(),
+                }),
+            ));
+
+            let mut stream = sql.query(&mut client).await?;
+            while let Some(item) = stream.try_next().await? {
+                match item {
+                    QueryItem::Metadata(meta) => {
+                        for col in meta.columns() {
+                            column_descs.push(ColumnDesc::named(
+                                col.name(),
+                                ColumnId::placeholder(),
+                                type_to_rw_type(&col.column_type())?,
+                            ));
+                        }
+                    }
+                    QueryItem::Row(row) => {
+                        unreachable!("Unexpected row: {:?}, `SELECT * FROM {} WHERE 1 = 0` should never return rows", row, config.table.clone());
+                    }
+                }
+            }
+        }
+        {
+            let sql = Query::new(format!(
```

Comment:
oh, you also want to support the `create table t (*)` syntax to read upstream table schema. I suggest you can split this feature to another pr in future.

Reply:
We can keep this impl I'll add test cases in the next PR.